Userland/Libraries/LibCore/Stream.h

```diff
@@ -663,7 +646,48 @@ class BufferedHelper {
             }
         }
 
-        auto readable_size = min(m_buffered_size, buffer.size());
+        auto readable_size = min(m_buffer.size(), buffer.size());
+
+        // As the underlying buffer is a CircularBuffer, we can't assume a
+        // continuous span of data. Therefore, we need to look at each
+        // distinct span.
+
+        struct SpanAndOffset {
+            ReadonlyBytes span {};
+            size_t offset {};
+        };
+
+        Vector<SpanAndOffset> spans_and_offsets;
+        Optional<ByteBuffer> overlapping_buffer;
+        auto const buffer_spans = m_buffer.each_separate_span();
+        if (readable_size <= buffer_spans[0].size()) {
+            // There is enough data, the span is continuous
+            spans_and_offsets.empend(buffer_spans[0].slice(0, readable_size), 0ul);
+        } else {
+            // We need to look through both spans and also check for edge
+            // cases where the candidate could overlap both span.
+
+            // The first buffer, which ends the underlying buffer.
+            spans_and_offsets.empend(buffer_spans[0], 0ul);
+
+            Optional<size_t> longest_candidate;
+            for (auto& candidate : candidates)
+                longest_candidate = max(candidate.length(), longest_candidate.value_or(candidate.length()));
+
+            // The second buffer, that overlaps the two others
+            auto const first_buffer_slice_size = min(*longest_candidate, buffer_spans[0].size());
+            auto const second_buffer_slice_size = min(min(*longest_candidate, buffer_spans[1].size()), readable_size - buffer_spans[0].size());
+
+            overlapping_buffer = TRY(ByteBuffer::create_uninitialized(first_buffer_slice_size + second_buffer_slice_size));
+
+            overlapping_buffer->overwrite(0, buffer_spans[0].slice_from_end(first_buffer_slice_size));
+            overlapping_buffer->overwrite(first_buffer_slice_size, buffer_spans[1].slice(0, second_buffer_slice_size));
+
+            spans_and_offsets.empend(overlapping_buffer->span(), buffer_spans[0].size() - first_buffer_slice_size);
+
+            // The third and final buffer, that begins the underlying buffer.
+            spans_and_offsets.empend(buffer_spans[1].slice(0, min(buffer_spans[1].size(), readable_size - buffer_spans[0].size())), buffer_spans[0].size());
+        }
```

Comment:
This looks like a lot of work just to determine if we at some point find a symbol that we want. Since the end result is to just fill the buffer anyways if we don't find the symbol, how about we just optimistically read data into the buffer (in some arbitrarily chosen chunk size) and add a function to the `CircularBuffer` to push back data to the beginning of the read queue in case we did find the delimiter?

Either that or we just implement the search internally, with a bit of delimiter splitting to work around the edges, because exposing internals like this isn't particularly great.

Reply:
Hmm, I'm really not fond of pushing data to the beginning of the read queue. And I suppose that it won't be fun to manage cases where people call write in the meantime.
What about adding a "`const` read", so reading again will send back the same data, and call discard if we found something.

Reply:
In fact, `memmem` has an overload that takes iterators that's probably the best fit.
We lose optimisations of `memmem_optional`, but that's still better than copying the data to iterate over it.

Reply:
Still not particularly great in my opinion, we should really be looking into an internal implementation (`Optional<size_t> offset_of(ReadonlyBytes needle)`?) then.

As for the suggestion with a `const` read, we should probably name it `peek` then, which is the usual terminology for "read but don't consume". Although, I can't see how one would ergonomically search through memory chunks using that.

Reply:
> And I suppose that it won't be fun to manage cases where people call write in the meantime.

`write` would append at the end anyways, the real 'problem' would be a `read` at the wrong time. However, I don't think that a multi-user circular buffer is the target at the moment either way?

Reply:
>Still not particularly great in my opinion

Adding iterators? And how would you implement `offset_of` then? That's basically reimplementing `KMP`

Reply:
> `write` would append at the end anyways

Sure, my bad I didn't explain my thought. I was thinking of filling the buffer so much that the `push_on_read` would override it.

Reply:
> Adding iterators?

"not particularly great" was not necessarily about how it works (I'm fine with basically anything as long as it's maintainable and not too spaghetti-like), but your message sounded like you wanted to make a public interface and let the caller handle what it actually wants to do with those iterators. If those are non-public I'm slightly more fond of the suggestion. :^)

That said, the existing `offset_off` implementations that I have seen that tried to take advantage of the memmem iterator API ended up having to jump through more hoops than it would have been to just do it manually (or they just forgot about the "what about the cutoff point" part completely).

> And how would you implement offset_of then?

I have drafted up an implementation of it for `AllocatingMemoryStream` [here](https://github.com/timschumi/serenity/commit/00081e7e493433b6b857e8fbbcddca29fb819b76). The only difference between that and the CircularBuffer is that we would only have to do this for one wraparound point instead of for multiple chunks. Obviously, for the range where we don't have to split the needle, we could just delegate to the standard `memmem`, but that's one for the future.

> That's basically reimplementing `KMP`

I'm not familiar with what `KMP` means in this context.

> I was thinking of filling the buffer so much that the push_on_read would override it.

We'd presumably just return an error as usual. However, I'd expect for "`unread`" calls to happen right after the respective `read`, so the buffer wouldn't even have the chance to fill enough to make the reverse operation fail.

Reply:
I finally implemented `offset_of`, I didn't understand how the `memset` overload with iterator work. So it wasn't has straightforward as I thought.
However, it is easy enough to use to implement `offset_of` in few lines.

> I have drafted up an implementation of it for `AllocatingMemoryStream` [here](https://github.com/timschumi/serenity/commit/00081e7e493433b6b857e8fbbcddca29fb819b76).

Have a look at the `memset` overload, AFAICT, it's exactly what you are looking for.

> I'm not familiar with what KMP means in this context.

It's the name of the algorithm used to implement `memmem` ([wikipedia](https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm)). And the reason why I think we should avoid implementing `offset_of` manually.